### PR TITLE
Add csi node user-facing configuration

### PR DESCRIPTION
**Pull request by copejon**
**Which issue(s) this PR addresses**:

Closes [USHIFT-168](https://issues.redhat.com//browse/USHIFT-168)

### CODE REVIEW

**Comment by ggiguash** [comment]
@copejon - could we reorganize the commits into at most 3?

File: pkg/components/components.go
Comment (ggiguash) [inline pkg/components/components.go:16]: Use filepath.join here?

**Comment by oglok** [comment]
Just not to forget, please, squash the commits if possible to avoid having 25 commits in history.

File: pkg/config/storage/lvmd.go
Comment (dhellmann) [inline pkg/config/storage/lvmd.go:16]: Another safer way to do this would be to add a "device classes" field to the MicroShift configuration file. That avoids exposing the `socket-name` option, so the user can't break it, but gives them control over the values they need to be able to change. It also lets us validate those settings, so we could fail to start if there is a list of device classes but none has the name "default", for example.

**Review by dhellmann — COMMENTED** [review]
I'm a little worried about the depth at which we're exposing the config options here. I'm also not that keen on having our config contain embedded YAML (we can't do much about the lvmd.yaml spec, but we can avoid that pattern ourselves).

File: pkg/config/storage/lvmd.go
Comment (copejon) [reply pkg/config/storage/lvmd.go:16]: I'd done that initially in fact but there were a few things that bugged me with the approach:

- The lvmd process already validates the config, concisely logs validation errors, and crashes immediately.  So it's obvious that, when the lvmd pod is crashLooping, `oc logs` will produce less than 5 lines describing what's wrong.  If microshift were made to crash on lvmd validation errors, the cause is much less obvious with the amount of log spam the process generates.  

- copying the lvmd config API in microshift feels like bad practice, and it also implies reflecting how lvmd validates the config fields.  The lvmd version is pinned in microshift of course, but it would essentially create an invisible dependency on the lvmd code base.  Less horrible that making lvmd a direct dependency, admittedly.

- Users can view the lvmd configmap in the cluster, edit it, and test config changes live against the cluster.  It ends up being a much cleaner debugging approach.

File: pkg/components/render_test.go
Comment (dhellmann) [inline pkg/components/render_test.go:68]: https://github.com/stretchr/testify has some nice features for showing where differences are in string comparisons like this. In another PR we should look at what adding that would do to the dependency list.

File: pkg/config/lvmd/lvmd.go
Comment (dhellmann) [inline pkg/config/lvmd/lvmd.go:33]: Should this be a pointer so we can tell the difference between being willing to not leave any spare disk and not providing a value at all? Or does the value have to be > 0 to be valid?

File: pkg/config/config.go
Comment (dhellmann) [inline pkg/config/config.go:79]: Would it make sense to be more specific here and make the public name something like `odfLVM`? That gives us the ability to use a different name for a different plugin upstream, for example.

It's in the config file, so I think we can leave "config" out of the field name either way.

File: pkg/config/lvmd/lvmd.go
Comment (dhellmann) [inline pkg/config/lvmd/lvmd.go:66]: Would it make sense to use the validation in the vendored code instead of reproducing it? Or does that drag in a bunch of other dependencies?

File: pkg/components/render_test.go
Comment (copejon) [reply pkg/components/render_test.go:68]: I like testify, but yeah, I didn't want to stack on more dependencies for this PR.

File: pkg/config/lvmd/lvmd.go
Comment (copejon) [reply pkg/config/lvmd/lvmd.go:33]: Lvmd will default a null `spare-gb` field to 10Gb, so users will have explicitly define the spare-gb field for any other value.

File: pkg/config/config.go
Comment (copejon) [reply pkg/config/config.go:79]: Good point, will fix.

File: pkg/config/lvmd/lvmd.go
Comment (dhellmann) [reply pkg/config/lvmd/lvmd.go:34]: How does that work? With the struct the way it is, I would expect the value of `SpareGB` to be `0` if the user does not provide anything. Then when we render the struct, I would expect to see the explicit `0` passed it to lvmd. Does lvmd convert the 0 to 10?

**Review by dhellmann — COMMENTED** [review]
Sorry for nitpicking on naming. I do have one other question about the exposed data structure.

File: pkg/config/lvmd/lvmd.go
Comment (copejon) [reply pkg/config/lvmd/lvmd.go:34]: In lvmd, non-nil, non-negative values are valid inputs, and a nil value results in a 10gb default being used. IMO that seems overly cautious.  In prod, I would expect the system storage to remain static after the device is imaged, with the unallocated VG space being managed by lvmd.  Meaning the spared space would never be utilized. If that's a valid assumption, then defaulting to 0 seems reasonable.

File: pkg/components/render_test.go
Comment (dhellmann) [inline pkg/components/render_test.go:23]: These tests look right, but I'd feel more comfortable with them if the expected values were spelled out by hand in the loop below, rather than using a duplicate template and rendering the same structures again in a different way.

**Review by dhellmann — COMMENTED** [review]
This version looks good. I have one nit that can come in another PR if you want. I do think squashing that typo fix into the original commit would be good before we merge the PR.

File: pkg/components/render_test.go
Comment (copejon) [reply pkg/components/render_test.go]: Fixed! :)

**Comment by dhellmann** [comment]
I'm happy with this version. I'd like to give @fzdarsky a chance to comment on the configuration layout and whether we want to put the storage-related settings in a subsection.

/lgtm
/hold

**Comment by ggiguash** [comment]
I have two questions:
* Some files under controllers/, pkg/mdns and pkg/util contain tabulation changes in the license text. Do we want these in the current PR?
* Do we need to have user documentation updated to reflect this new functionality?

**Comment by copejon** [comment]
> I have two questions:
> 
>     * Some files under controllers/, pkg/mdns and pkg/util contain tabulation changes in the license text. Do we want these in the current PR?

It seems that it won't pass the verify check otherwise.  Not sure why this is being caught now though.

>     * Do we need to have user documentation updated to reflect this new functionality?

I'll circle back and make sure the in-repo docs are accurate.

File: docs/default_csi_plugin.md
Comment (pmtk) [inline docs/default_csi_plugin.md:34]:
```suggestion
The user provided lvmd config should be written to the same directory as the MicroShift config.  If a MicroShift config
```

File: docs/default_csi_plugin.md
Comment (pmtk) [inline docs/default_csi_plugin.md:39]:
```suggestion
2. Global config dir: `/etc/microshift/lvmd.yaml`
```

File: pkg/components/storage.go
Comment (dhellmann) [inline pkg/components/storage.go:69]: Renaming `l` to something like `lvmdSettings` here would make it more obvious what it holds.

File: pkg/config/lvmd/lvmd_test.go
Comment (dhellmann) [inline pkg/config/lvmd/lvmd_test.go:61]: Do we need to populate the `Lvmd` instance here in order to get valid defaults? It looks like `withDefaults` always returns the same data.

File: pkg/config/lvmd/types.go
Comment (pmtk) [inline pkg/config/lvmd/types.go:2]: Nit: We could add an empty line between doc and `package`. Otherwise godoc will treat the comment like doc for the package

File: pkg/config/lvmd/test/config.yaml
Comment (dhellmann) [inline pkg/config/lvmd/test/config.yaml:2]: I think we want to remove this line, right?

File: scripts/tools.go
Comment (pmtk) [inline scripts/tools.go:1]: so this makes sure that this package does not end up in the binary?

File: pkg/config/lvmd/test/lvmd.yaml
Comment (pmtk) [inline pkg/config/lvmd/test/lvmd.yaml:10]: nit: missing endline

File: pkg/config/lvmd/test/valueMismatch.yaml
Comment (pmtk) [inline pkg/config/lvmd/test/valueMismatch.yaml:10]: nit: missing endline

File: docs/default_csi_plugin.md
Comment (mangelajo) [inline docs/default_csi_plugin.md:39]:
```suggestion
2. Global config dir; `/etc/microshift/lvmd.yaml`
```

File: test/config.yaml
Comment (dhellmann) [inline test/config.yaml:4]: I think we want to remove this line, right?

**Review by mangelajo — CHANGES_REQUESTED** [review]
a tiny doc nit

File: pkg/config/lvmd/lvmd_test.go
Comment (dhellmann) [inline pkg/config/lvmd/lvmd_test.go:122]: What error is expected? Maybe you could add a comment to the file for now and then in a later PR update the test to check the error type or message?

File: scripts/tools.go
Comment (copejon) [reply scripts/tools.go:1]: Not exactly.  These tool-import files are a bit of a hack to get go mod to fetch executables.  Unfortunately, `go test` can recognize this, but will still fail.  So slapping an arbitrary build constraint is an idiomatic way to enable `go test ./...`.  I meant to remove this one however, since it was just to make my IDE happy.

Comment (copejon) [inline scripts/tools.go:1]:
```suggestion
```

File: test/config.yaml
Comment (copejon) [reply test/config.yaml:4]: Ah, yes.  It's ignored in the tests since the field doesn't exist on the go side. Will fix.

File: docs/default_csi_plugin.md
Comment (copejon) [reply docs/default_csi_plugin.md:35]:
```suggestion
doesn't exist, MicroShift will assume default lvmd values. These paths will be checked for the config, depending on the user MicroShift
```

File: pkg/config/lvmd/types.go
Comment (copejon) [reply pkg/config/lvmd/types.go:2]:
```suggestion
// https://github.com/red-hat-storage/topolvm/blob/release-4.10/lvmd/device_class_manager.go

```

File: pkg/config/lvmd/lvmd_test.go
Comment (copejon) [reply pkg/config/lvmd/lvmd_test.go]: I've removed this test.  withDefaults returns a static type, which would be checked at compile time.  No point in testing for it as well.

File: pkg/config/lvmd/lvmd_test.go
Comment (copejon) [reply pkg/config/lvmd/lvmd_test.go]: I've gone ahead and removed the test.

**Review by dhellmann — COMMENTED** [review]
/lgtm

**Pull request merged**
